File: python/problem121.py
# Time Limit Exceeded
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """

        ma=0

        for i in range(0,len(prices)):
            for k in range(i+1,len(prices)):
                if prices[k]-prices[i]>ma:
                    ma=prices[k]-prices[i]
        return ma


# from solution
class Solution2:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        try:
            mi=prices[0]
            ma=0

            for i in range(0,len(prices)):
                if prices[i]<mi:
                    mi=prices[i]


                elif prices[i]-mi>ma:
                        ma=prices[i]-mi
            return ma
        except:
            return 0      # input:[]  

File: swift/problem121.swift
class Solution1 {
    func maxProfit(_ prices: [Int]) -> Int {
        var min = Int.max; var profit = 0
        
        for price in prices {
            if (price < min) {
                min = price
            } else {
                profit = max(profit, price - min)
            }
        }
        
        return profit
    }
}

//sorting the price together with date
class Solution2 {
    func maxProfit(_ prices: [Int]) -> Int {
        let length = prices.count
        let prices = prices.enumerated().sorted{$0.element < $1.element}
        var res = 0; var prev = Int.max
        
        for i in 0..<length {
            var j = length - 1
            if (prices[i].offset > prev) {
                prev = prices[i].offset
                continue
            }
            prev = prices[i].offset
            while (j > i) {
                if (prev < prices[j].offset) {
                    res = max(prices[j].element - prices[i].element, res)
                    break
                }
                j -= 1
            }
        }
        
        return res
    }
}
